nimbus/src/enrollment.rs

```diff
@@ -85,90 +365,156 @@ pub fn get_enrollments(db: &Database) -> Result<Vec<EnrolledExperiment>> {
+        let enrollments_store = db.get_store(StoreId::Enrollments);
+        let existing_experiments = experiments_store.collect_all::<Experiment>(writer)?;
+        let existing_enrollments =
+            enrollments_store.collect_all::<ExperimentEnrollment>(&writer)?;
+        // Calculate the changes.
+        let (updated_enrollments, enrollments_change_events) = self.evolve_enrollments(
+            is_user_participating,
+            &existing_experiments,
+            updated_experiments,
+            &existing_enrollments,
+        )?;
+        // Write the changes to the Database.
+        enrollments_store.clear(writer)?;
+        for enrollment in updated_enrollments {
+            enrollments_store.put(writer, &enrollment.slug, &enrollment)?;
+        }
+        experiments_store.clear(writer)?;
+        for experiment in updated_experiments {
+            experiments_store.put(writer, &experiment.slug, experiment)?;
+        }
+        Ok(enrollments_change_events)
+    }
+
+    /// Evolve and calculate the new set of enrollments, using the
+    /// previous and current state of experiments and current enrollments.
+    pub(crate) fn evolve_enrollments(
+        &self,
+        is_user_participating: bool,
+        existing_experiments: &[Experiment],
+        updated_experiments: &[Experiment],
+        existing_enrollments: &[ExperimentEnrollment],
+    ) -> Result<(Vec<ExperimentEnrollment>, Vec<EnrollmentChangeEvent>)> {
+        let mut updated_enrollments = vec![];
+        let mut enrollment_events = vec![];
+        let existing_experiments = map_experiments(&existing_experiments);
+        let updated_experiments = map_experiments(&updated_experiments);
+        let existing_enrollments = map_enrollments(&existing_enrollments);
+
+        let mut all_slugs = HashSet::with_capacity(existing_experiments.len());
+        all_slugs.extend(existing_experiments.keys());
+        all_slugs.extend(updated_experiments.keys());
+        all_slugs.extend(existing_enrollments.keys());
+
+        for slug in all_slugs {
+            let updated_enrollment = self.evolve_enrollment(
+                is_user_participating,
+                existing_experiments.get(slug).copied(),
+                updated_experiments.get(slug).copied(),
+                existing_enrollments.get(slug).copied(),
+                &mut enrollment_events,
+            )?;
+            if let Some(enrollment) = updated_enrollment {
+                updated_enrollments.push(enrollment);
+            }
+        }
+
+        Ok((updated_enrollments, enrollment_events))
+    }
+
+    fn evolve_enrollment(
+        &self,
+        is_user_participating: bool,
+        existing_experiment: Option<&Experiment>,
+        updated_experiment: Option<&Experiment>,
+        existing_enrollment: Option<&ExperimentEnrollment>,
+        enrollment_events: &mut Vec<EnrollmentChangeEvent>,
+    ) -> Result<Option<ExperimentEnrollment>> {
```

Comment:
I like the abstraction level of `evolve_enrollment` here, it feels like the sort of thing for which we could have a bit old suite of tests for all the various combos of things that change on the server and assert that the appropriate change happens on the client.